src/odemis/driver/keysight.py

```diff
@@ -122,144 +187,259 @@ def _sendCommand(self, cmd):
         """
-        self._sendCommand(b"sour%d:phas %e sec" % (c, t))
-        # self._sendCommand(b"phas 1e-6 SEC")  #DEBUG
+        self._sendCommand(b"sour%d:phas %.15e sec" % (c, t))
+        self._checkError()
 
-    def setDutyCycle(self, c, dc):
+    def getPhase(self, c: int) -> float:
         """
-        c (int, 1 or 2): the channel to set the duty cycle of.
-        dc (float): the duty cycle percentage. between 00.01 and 99.99
+        c (1 or 2): the channel to get the phase of.
+        :return: the phase value in degrees (-360 to 360)
         """
-        self._sendCommand(b"sour%d:func:squ:dcyc %.2f" % (c, dc))
+        ans = self._sendQuery(b"sour%d:phas?" % c)
+        return float(ans)
+
+    def setDutyCycle(self, c: int, dc: float):
+        """
+        c (1 or 2): the channel to set the duty cycle of.
+        dc: the duty cycle percentage. between 00.01 and 99.99, or smaller values if the frequency is high
+        :raise: TrueFormError if the duty cycle is out of bounds. The device will
+        typically clip the value to the nearest valid value.
+        """
+        self._sendCommand(b"sour%d:func:squ:dcyc %.4f" % (c, dc))
+        try:
+            self._checkError()
+        except TrueFormError as e:
+            if e.errno == -222:  # Data out of range
+                raise IndexError(f"Duty cycle {dc} out of range: {e.strerror}")
+
+    def getDutyCycle(self, c: int) -> float:
+        """
+        c (1 or 2): the channel to get the duty cycle of.
+        :return: the duty cycle percentage
+        """
+        ans = self._sendQuery(b"sour%d:func:squ:dcyc?" % c)
+        return float(ans)
 
     def setTracking(self, c: int, t: str) -> None:
         """
         c (1 or 2): the channel to set the tracking mode of.
         t (ON, OFF or INV): the tracking mode
         """
-        if t not in ["ON", "OFF", "INV"]:
+        if t.upper() not in ["ON", "OFF", "INV"]:
             raise ValueError("Incorrect tracking mode: %s given" % (t,))
+
         self._sendCommand(b"sour%d:trac %s" % (c, t.encode("ascii")))
+        self._checkError()
 
-    def setWaveform(self, c, f):
+    def setWaveform(self, c: int, f: str):
         """
-        c (int, 1 or 2): the channel to set the tracking mode of.
-        f (str SIN, SQU, RAMP or PULS): the wave form to generate
+        c (1 or 2): the channel to set
+        f (SIN, SQU, RAMP or PULS): the waveform to generate
         """
-        valid_waveforms = {b"SIN", b"SQU", b"RAMP", b"PULS"}
-        if f.upper() not in (waveform.upper() for waveform in valid_waveforms):
+        if f.upper() not in {"SIN", "SQU", "RAMP", "PULS"}:
             raise ValueError("Incorrect waveform: %s given" % (f,))
 
-        self._sendCommand(b"sour%d:appl:%s" % (c, f))
+        self._sendCommand(b"sour%d:appl:%s" % (c, f.encode("ascii")))
+        self._checkError()
 
     def _set_period(self, p: float) -> float:
-        f = 1/p
+        f = 1 / p
```

Comment:
To be safe, can we make sure p>0

Reply:
It's already taken care of by the VA. I'll add comment.